### PR TITLE
feat(mechanics): NPCs can have actions for various ShipEvents

### PR DESCRIPTION
**Pull request by warp-core**
**Feature:** This PR implements the feature request detailed and discussed in issue #5248, and probably some others.

## Feature Details
Mission NPCs can have `MissionAction`s triggered by `ShipEvent`s occurring to their ships.

This is super incomplete, but it should be possible to test it at a basic level.

This will probably end up with a new NPCAction class instead of using MissionAction, since I expect the stuff we might want to do with this will extend well beyond what MissionAction supports. Stuff like jumping into a system with the player's ship, or a system more than <x> jumps away from the player's ship (fail escort if the ship gets too far away from you???)

This could also likely be expanded to support Zitchas' request for ships to change personality or government after being boarded by a certain faction, though I think that would require some other changes which I don't want to do in this PR.

## Usage Examples
```
npc
  on (assist | scan cargo | scan outfits | provoke | disable | board | capture | destroy | atrocity)
    <anything you can put in a MissionAction>
```

## Testing Done
I'm surprised it even built tbh.

Gave a mission npc with the `kill` objective an `on disable` conversation and get that conversation when it is disabled.

## Performance Impact
N/A?

### CODE REVIEW

File: source/NPC.cpp [hunk 1]
```diff
@@ -444,10 +451,15 @@ void NPC::Do(const ShipEvent &event, PlayerInfo &player, UI *ui, bool isVisible)
...
 	// Apply this event to the ship and any ships it is carrying.
-	actions[ship.get()] |= type;
+	shipActions[ship.get()] |= type;
 	for(const Ship::Bay &bay : ship->Bays())
 		if(bay.ship)
-			actions[bay.ship.get()] |= type;
+			shipActions[bay.ship.get()] |= type;
+
+	string typeString = ShipEvent::TypeToString(type);
+	auto it = npcActions.find(typeString);
```
Comment (RisingLeaf) [inline source/NPC.cpp:460]:
```suggestion
	auto it = npcActions.find(ShipEvent::TypeToString(type));
```
If it is only used once it is not really worth another variable, is it?

File: source/ShipEvent.cpp [hunk 2]
```diff
@@ -74,3 +74,25 @@ int ShipEvent::Type() const
 {
 	return type;
 }
+
+
+
+string ShipEvent::TypeToString(int type)
+{
+	static const map<int, string> types = {
```
Comment (RisingLeaf) [inline source/ShipEvent.cpp:82]: This map does not need to be redefined every function cast.
How about storing it somewhere else?

File: source/MissionAction.cpp [hunk 3]
```diff
@@ -110,7 +111,7 @@ void MissionAction::Load(const DataNode &node, const string &missionName)
 				child.PrintTrace("Error: Skipping invalid \"require\" amount:");
 		}
 		// The legacy syntax "outfit <outfit> 0" means "the player must have this outfit installed."
```
Comment (RisingLeaf) [inline source/MissionAction.cpp:113]: Maybe you want to add the npc condition to the comment here?

**Comment by RisingLeaf** [comment]
I know it is only draft but I thought there is now harm in looking over it.

File: source/ShipEvent.cpp (hunk 2 above)
Comment (warp-core) [reply source/ShipEvent.cpp:82]: I'm fairly sure it won't be reconstructed/reallocated every time.
It is static and I think this method is static, so it'll either always exist or be constructed on first use.
Putting it in here just scopes it to this method only. If it is used anywhere else, it can be moved to a namespace or the class or something.

Comment (Hurleveur) [reply source/ShipEvent.cpp:82]: Indeed static const will only be defined the first time

**Comment by Hurleveur** [comment]
I need this for the Unfettered campaign, so do tell me whenever its ready to be looked at and I'll review it.

**Comment by warp-core** [comment]
That wasn't as bad as I expected.

~~inb4 doesn't build~~

File: source/MissionAction.cpp [hunk 4]
```diff
@@ -112,6 +112,12 @@ void MissionAction::Load(const DataNode &node, const string &missionName)
...
 		else if(key == "outfit" && child.Size() >= 3 && child.Token(2) == "0")
 		{
+			// Do not support this syntax for NPC actions, since they were created after it was deprecated.
+			if(isNPC)
```
Comment (Hurleveur) [inline source/MissionAction.cpp:116]: shouldnt it be `inNPC` or `forNPC` or `ìnNPCDefinition`?

or mb `ìsNPCAction`

File: source/NPC.cpp [hunk 5]
```diff
@@ -151,6 +152,16 @@ void NPC::Load(const DataNode &node)
 			else
 				child.PrintTrace("Skipping unrecognized attribute:");
 		}
+		else if(child.Token(0) == "on" && child.Size() >= 2)
+		{
+			int eventType = ShipEvent::TypeFromString(child.Token(1));
```
Comment (Hurleveur) [inline source/NPC.cpp:157]: we def should be using this whole method at the start of this function as well, within a lambda but that could be for another PR

File: source/NPC.h [hunk 6]
```diff
@@ -56,9 +57,9 @@ class NPC {
 	~NPC() noexcept = default;
 
 	// Construct and Load() at the same time.
-	NPC(const DataNode &node);
+	NPC(const DataNode &node, std::string missionName = "");
```
Comment (Hurleveur) [inline source/NPC.h:60]: do we need a default argument? I feel like we always give this missionName, dont we?

File: source/NPC.cpp [hunk 7]
```diff
@@ -151,6 +152,16 @@ void NPC::Load(const DataNode &node)
...
+			if(eventType == ShipEvent::JUMP)
+				child.PrintTrace("Error: \"jump\" is not a supported event for NPC actions.");
+			else if(eventType == ShipEvent::NONE)
+				child.PrintTrace("Error: invalid ShipEvent.");
+			else
+				npcActions.insert(make_pair(eventType, MissionAction(child, missionName, true)));
```
Comment (Hurleveur) [inline source/NPC.cpp:163]: why do we need to make pair? we're inserting into a map, right?

**Review by Hurleveur — COMMENTED** [review]
looking good, I like those new methods to parse the ShipEvents

File: source/NPC.cpp (hunk 5 above)
Comment (warp-core) [reply source/NPC.cpp:157]: The beginning of `Load` does a very different thing to this.

Comment (Hurleveur) [reply source/NPC.cpp:157]: Oh right its parsing die save etc? Why would we use different names here?

Comment (warp-core) [reply source/NPC.cpp:157]: Supporting actions triggered by individual ship events is fairly simple, more complex behaviours like hte objectives is more complicated.
What would it mean to have an `on save` NPC action? Complete this mission with the whole NPC alive? Does the NPC get told when we reach the destination? Is some event sent at that time?
Potential for extension there, but it seems excessive, and not necessary if actions can be triggered by ship events already.

File: source/MissionAction.cpp [hunk 8]
```diff
@@ -112,6 +112,12 @@ void MissionAction::Load(const DataNode &node, const string &missionName)
 		// The legacy syntax "outfit <outfit> 0" means "the player must have this outfit installed."
 		else if(key == "outfit" && child.Size() >= 3 && child.Token(2) == "0")
 		{
+			// Do not support this syntax for NPC actions, since they were created after it was deprecated.
+			if(isNPCAction)
+			{
+				child.PrintTrace("Error: unsupported use of \"outfit\" inside an NPC action.");
+				continue;
+			}
```
Comment (Amazinite) [inline source/MissionAction.cpp:120]: It doesn't seem like there's any actual harm done in allowing this action to be used on an NPC, so I'd just remove this warning.

**Review by Amazinite — CHANGES_REQUESTED** [review]

**Comment by Amazinite** [comment]
Took another look at this and realized I wasn't too big a fan of the approach that was taken. By tying NPC mission actions directly to ship events, we could only ever have a one-to-one relationship between ship events and mission actions that they trigger. This would prevent some of the ideas mentioned in #5248 from being done, or at least being done easily. The way this was implemented also meant that *every* ship event would trigger the associated action for every time that it was done. For example, a fleet of multiple ships with an `on disable` action would have ran that action for each ship that became disabled.

I've changed this PR so that how an NPC handles mission actions is closer to how a Mission handles them, by using a Trigger enum that defines "trigger points" for an NPC's actions. I've mapped various ShipEvents to these Triggers, and constructed the mapping in such a way that we could have the same ShipEvent potentially invoke multiple Triggers.
```c++
// Map the ShipEvent that was received to the Triggers it could flip.
static const map<int, vector<Trigger>> eventTriggers = {
	{ShipEvent::DESTROY, {KILL}},
	{ShipEvent::BOARD, {BOARD}},
	{ShipEvent::ASSIST, {ASSIST}},
	{ShipEvent::DISABLE, {DISABLE}},
	{ShipEvent::SCAN_CARGO, {SCAN_CARGO}},
	{ShipEvent::SCAN_OUTFITS, {SCAN_OUTFITS}},
	{ShipEvent::CAPTURE, {CAPTURE}},
	{ShipEvent::PROVOKE, {PROVOKE}},
};
```
The behavior of the existing Triggers has also been changed to match the behavior of their associated objective. That is, run once every ship in the fleet has been hit by a given ShipEvent instead of running on each ship, although that behavior isn't strictly enforced, and future Triggers could run their actions under different scenarios.

By using Triggers as an intermediary between ShipEvents and MissionActions, we also open the possibility of creating NPC actions that aren't based on ShipEvents, which also better supports some of the ideas mentioned in #5248.

**Comment by Amazinite** [comment]
Tested each NPC action with a fleet of one ship and a fleet of two ships. 
* `on disabled` triggers once all ships in the fleet have been disabled. Can retrigger if a ship is assisted then disabled again.
* `on kill` triggers twice, as the destroy event is sent once when a ship's HP hits zero, and a second time once it explodes. The second time is intended to capture self-destructing ships. Should make NPC only handle this action once.
* `on board` triggers once all ships in the fleet have been boarded, then retriggers on repeat boardings.
* `on assist` triggers once all ships in the fleet have been assisted, then retriggers on repeat assists.
* `on "scan cargo"` and `on "scan outfits"` trigger once all ships in the fleet have been scanned, then retrigger on repeat scans.
* `on capture` triggers once all ships in the fleet have been captured.
* `on provoke` triggers a variable number of times (I got from one to four triggers on the same ship after provoking it), only on single-ship fleets, and only if the ship from the NPC is the one who was provoked. That is, provoking another ship from the NPC's government won't trigger the provoke action.

So the changes to make are:
* Don't double trigger `on kill` for ships which are destroyed and then explode.
* Allow `on provoke` to work for multi-ship fleets.
* Find out why the provoke ship event is sending repeatedly, and either only send it once or only trigger the provoke action once.

I'll work on these later.

**Comment by Amazinite** [comment]
Didn't find out why provoke events are repeatedly sent, but it's a moot point given single use actions. If ever we wanted a repeatable provoke action, we'd need to look into that.
